### PR TITLE
Raise an error in np.var when the input array is complex and dtype is not

### LINKED ISSUE #2283: np.var does not match onp.var in some cases
Running this snippet
```python
import jax.numpy as np 
import numpy as onp 
inp_array = np.array([[[5.29215704-0.30965656j,  1.20047163+1.23179551j, 0+0.43213071j,  6.7226796 +4.36282052j]],   
                      [[5.60267397+2.28311318j,           0+0.36502505j,         0 +1j, -0.45407162+1.00102298j]]])  
print(np.var(inp_array, 2, dtype=np.float32)) 
print(onp.var(onp.array(inp_array), 2, dtype=onp.float32))
```
Prints different outputs for numpy and jax.numpy
```
[[10.911064]
 [ 6.728332]]
[[12.953884]
 [ 8.079251]]
```
Not sure if there is an actual use case for np.var on complex arrays with dtype=float, but I guess the outputs are supposed to be the same since jax.numpy is supposed to match numpy where it's possible.

The mismatch happens because the implementation in NumPy uses dtype for all the calculations, meanwhile, jax.numpy only converts the computed value to dtype in the very end and does all the calculations in the promote_types(complex, dtype=float) which evaluates to complex.

If this sounds like something that should be fixed then I can make a PR.

### PR DESCRIPTION
**Pull request by yurodiviy**
fixes #2283

### CODE REVIEW

**Comment by yurodiviy** [comment]
@googlebot I fixed it.

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -151,6 +151,8 @@ def _make_scalar_type(onp_scalar_type):
 complex64 = csingle = _make_scalar_type(onp.complex64)
 complex128 = cdouble = _make_scalar_type(onp.complex128)
 
+complex_dtypes = [complex64, complex128]
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: These variable names end up exported from `jax.numpy`, so we should avoid polluting the namespace. In general, and underscore prefix on the name will do the trick.

However, in this specific case, I think you might simply write:
`issubdtype(dtype, complexfloating)`
?
You then don't need this definition.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -1399,6 +1401,9 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
 
   a_dtype = _dtype(a)
   if dtype:
+    if dtype not in complex_dtypes and a_dtype in complex_dtypes:
+      raise ValueError("dtype must be complex when calling var on a "
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Add a comment explaining the reasoning here, given NumPy doesn't error here?

**Review by hawkinsp — CHANGES_REQUESTED** [review]

File: jax/numpy/lax_numpy.py (hunk 1 above)
Comment (yurodiviy) [reply jax/numpy/lax_numpy.py]: Thanks for catching that, changed it to issubdtype.

File: jax/numpy/lax_numpy.py (hunk 2 above)
Comment (yurodiviy) [reply jax/numpy/lax_numpy.py]: Sure, added a sentence explaining why we're raising an error.

Comment (hawkinsp) [reply jax/numpy/lax_numpy.py]: The issue with this error is that it isn't actionable for the user.

Perhaps change the error to something like:

jax.numpy.var does not yet support real dtype parameters when computing the variance of an array of complex values. The semantics of numpy.var seem unclear in this case. Please comment on https://github.com/google/jax/issues/2283 if this behavior is important to you.

Comment (yurodiviy) [reply jax/numpy/lax_numpy.py]: Changed the error message as per your suggestion.

**Comment by yurodiviy** [comment]
@hawkinsp Any other changes that I need to make before we can merge this?

**Pull request merged**